oidc_middleware.go

```diff
@@ -0,0 +1,88 @@
+package main
+
+import (
+	"crypto/rsa"
+	"encoding/base64"
+	"encoding/json"
+	"github.com/gin-gonic/gin"
+	"math/big"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/golang-jwt/jwt"
+)
+
+// Largely taken from https://developer.okta.com/blog/2021/01/04/offline-jwt-validation-with-go
+
+func GetOidcPublicKeys(config OIDCConfig) *map[string]*rsa.PublicKey {
+	rsaKeys := make(map[string]*rsa.PublicKey)
+	var body map[string]interface{}
+	resp, err := http.Get(config.KeysURL)
+	if err != nil {
+		log.Error().WithError(err).Message("Could not fetch from KeysURL.")
+	}
+	err = json.NewDecoder(resp.Body).Decode(&body)
+	if err != nil {
+		log.Error().WithError(err).Message("Failed to decode login JWT Keys.")
+	}
+	log.Debug().Message("Updating keys for oidc.")
+	rsaExponent := 65537
+	for _, bodykey := range body["keys"].([]interface{}) {
+		key := bodykey.(map[string]interface{})
+		kid := key["kid"].(string)
+		rsakey := new(rsa.PublicKey)
+		number, _ := base64.RawURLEncoding.DecodeString(key["n"].(string))
+		rsakey.N = new(big.Int).SetBytes(number)
+		rsakey.E = rsaExponent
+		rsaKeys[kid] = rsakey
+	}
+	return &rsaKeys
+}
+
+func VerifyTokenMiddleware(config OIDCConfig, rsaKeys *map[string]*rsa.PublicKey) gin.HandlerFunc {
+	return func (ginContext *gin.Context) {
+		if *rsaKeys == nil {
+			log.Warn().Message("RsaKeys for OIDC have not been set (http:500).")
+			ginContext.AbortWithStatus(http.StatusInternalServerError)
+		}
+		isValid := false
+		errorMessage := ""
+		tokenString := ginContext.Request.Header.Get("Authorization")
+		if !strings.HasPrefix(tokenString, "Bearer ") {
+			ginContext.AbortWithStatus(http.StatusUnauthorized)
+		}
+		tokenString = strings.TrimPrefix(tokenString, "Bearer ")
+		token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
+			return (*rsaKeys)[token.Header["kid"].(string)], nil
+		})
```

Comment:
As @jilleJr mentioned earlier, better to make sure the key actually exists so it doesn't panic and we can handle the situation more gracefully.

If somebody would pass a properly formed token with `kid` header that we don't have registered then this would panic and respond with `500 Internal Server Error` (I think) basically telling us and the end-user nothing about what went wrong, while `401 Unauthorized` along with some info about what went wrong would be better for both.

Perhaps extracting it to something like:
```go
func tokenToRSAKey(token *jwt.Token) (interface{}, error) {
	kid, ok := token.Header["kid"]
	if !ok {
		return nil, errors.New("no header 'kid' found")
	}
	kidStr, ok := kid.(string)
	if !ok {
		return nil, erorrs.New("value of 'kid' header not convertible to string")
	}
	rsaKey, ok := (*rsaKeys)[kidStr]
	if !ok {
		return nil, errors.New("no matching RSA key found")
	}
	return rsaKey, nil
}
```

and then change to:

```suggestion
		tokenString = strings.TrimPrefix(tokenString, "Bearer ")
		token, err := jwt.Parse(tokenString, tokenToRSAKey)
		if err != nil {
			// Unsure about the detail message here, but something like that maybe?
			ginutil.WriteUnauthorizedError(c, err, "Unable to find matching RSA key for JWT token.")
			return
		}
```

? Increasing the brevity of the handler while helping with debugging/end-user experience along the line.

While a lot longer (could probably be made nicer though), it is only 17 lines that could save a fair amount of debugging time. Especially considering we don't really get any proper stacktraces.

It's a tradeoff for sure, but I personally prefer shortening debugging time, and providing better info to the end-user, more than fewer lines of code. If we can do both then that's golden.

Reply:
Dealt with panic-prone code.